File: backend/src/domain/percentile.ts
/**
 * Midpoint percentile rank: for a distribution of counts, returns how "high"
 * `count` scores as a percentage (0-100). Ties contribute half their mass so
 * "everyone tied at top" maps to 50 — more honest than the strict-below or
 * inclusive-at-or-below variants.
 */
export function percentileRankFor(
  count: number,
  distribution: Record<string, number>,
): number {
  let total = 0;
  let below = 0;
  let equal = 0;
  for (const [bucketRaw, n] of Object.entries(distribution)) {
    const bucket = Number(bucketRaw);
    if (!Number.isFinite(bucket) || n <= 0) continue;
    total += n;
    if (bucket < count) below += n;
    else if (bucket === count) equal += n;
  }
  if (total === 0) return 0;
  return Math.round(((below + equal / 2) / total) * 100);
}

File: mobile/lib/domain/percentile.dart
/// Midpoint percentile rank over a histogram {"count": users}: how "high"
/// [count] scores as 0–100. Ties contribute half their mass, so "everyone
/// tied" maps to 50 — more honest than strict-below.
int percentileRankFor(int count, Map<String, int> distribution) {
  var total = 0;
  var below = 0;
  var equal = 0;
  for (final entry in distribution.entries) {
    final bucket = int.tryParse(entry.key);
    final users = entry.value;
    if (bucket == null || users <= 0) continue;
    total += users;
    if (bucket < count) {
      below += users;
    } else if (bucket == count) {
      equal += users;
    }
  }
  if (total == 0) return 0;
  return ((below + equal / 2) / total * 100).round();
}

/// The free tier's world comparison: a sourced healthy range, fully offline.
/// ("Most people land between 10 and 20 a day.")
const int kWorldRangeLow = 10;
const int kWorldRangeHigh = 20;

enum WorldPace { quiet, onPace, breezy }

WorldPace worldPaceFor(int todayCount) {
  if (todayCount < kWorldRangeLow) return WorldPace.quiet;
  if (todayCount <= kWorldRangeHigh) return WorldPace.onPace;
  return WorldPace.breezy;
}
